shell-script for updating autodeploy xars on build time

bwbohl opened the pull request:
This adds a shell script that will be copied into the container and executed if the docker --build-arg UPDATE_XARS=true. Submitting the $VERSION it will fetch apps.xml from eXist-db's public repo and fetch the latest XAR-versions for this eXist version.

bwbohl committed:
add build ARG for updating autodeploy xars on build time

bwbohl committed:
Update XARs on docker build

add build ARG for updating autodeploy xars on build time

Update README.md

with information on new build-arg UPDATE_XARS

update xars on build

bwbohl committed:
add build ARG for updating autodeploy xars on build time

bwbohl committed:
Update XARs on docker build

add build ARG for updating autodeploy xars on build time

Update README.md

with information on new build-arg UPDATE_XARS

update xars on build

bwbohl commented:
open for discussion though still see some room for improvement ;-)

bwbohl committed:
Update README.md

with information on new build-arg UPDATE_XARS

bwbohl committed:
Update README.md

with information on new build-arg UPDATE_XARS

bwbohl commented:
at the moment the packages to be pulled are hard coded, would it be better to iterate through the contents in the auto deploy directory to see which packages are in autodeploy? My impression is though, that this is quite stable over all version of eXist-db, or am I mistaken?

peterstadler commented on Dockerfile:
```diff
@@ -12,6 +12,7 @@ RUN addgroup -S wegajetty \
 
 ARG VERSION
 ARG MAX_MEMORY
+ARG UPDATE_XARS
```
I think this is unnecessary. Why would you want to have old versions (as default)? IMHO this ARG can be removed completely.

bwbohl replied:
If you want to add some custom XARs (see below) it might be nice to not autoupdate them as the script at the moment flushes all contents of the auto deploy folder.

peterstadler commented on Dockerfile:
```diff
@@ -20,14 +21,15 @@ ENV MAX_MEMORY ${MAX_MEMORY:-2048}
 ENV EXIST_ENV ${EXIST_ENV:-development}
 ENV EXIST_CONTEXT_PATH ${EXIST_CONTEXT_PATH:-/exist}
 ENV EXIST_DATA_DIR ${EXIST_DATA_DIR:-/opt/exist/data}
+ENV UPDATE_XARS ${UPDATE_XARS:-false}
```
remove. see above

peterstadler commented on Dockerfile:
```diff
@@ -44,6 +46,14 @@ RUN apk --update add bash pwgen curl \
 
 # adding expath packages to the autodeploy directory
+COPY update-xars.sh /tmp/update-xars.sh
+RUN chmod +x /tmp/update-xars.sh
+RUN if [[ ${UPDATE_XARS} = "true" ]] ; \
+    then /tmp/update-xars.sh ${VERSION} /tmp/xar-updates/  \
+    && rm -f ${EXIST_HOME}/autodeploy/*.xar \
+    && cp /tmp/xar-updates/*.xar ${EXIST_HOME}/autodeploy/ \
+    && rm -Rf /tmp/xar-updates/ ; \
+    else echo will not update xars ; fi 
```
I think all this logic  is better moved into the `update-xars` script itself. Here it should only be 
```shell
RUN chmod +x /tmp/update-xars.sh && /tmp/update-xars.sh ${VERSION}
```

bwbohl replied:
good point

peterstadler commented on README.md:
```diff
@@ -18,6 +18,7 @@ docker build -t existdb --build-arg VERSION=5.2.0 .
 * **VERSION**: The eXist version to use. Defaults to 5.2.0
 * **MAX_MEMORY**: The max memory setting for eXist. Defaults to 2048 (megabyte). 
    This can be overwritten when starting a container by adding e.g. `-e JAVA_OPTIONS="-Xmx1024m"`    
+* **UPDATE_XARS**: If set to true, the contents of the autodeploy folder will be replaced with the latest package versions of dashboard, eXide, exist-documentation, exist-function-documentation, functx, markdown, monex, packageservice, semver-xq, and shared-resource available through the eXist-db public repo for the eXist-db specified with `VERSION`
```
remove. see above

peterstadler commented on update-xars.sh:
```diff
@@ -0,0 +1,49 @@
+VERSION=$1
+DIR=$2
```
I think this is unnecessary. Just create a tmp folder with `mktemp -d` and don't bother pushing this around as a parameter.

bwbohl replied:
good thing when pushing this around is that the script can be used standalone when preparing a instance with some additional custom XARs. This way you could run it before starting the db, add your custom XARs to the same folder and `-v YOUR-FOLDER:/opt/exist/autodeploy` when calling `docker run`

peterstadler commented on update-xars.sh:
```diff
@@ -0,0 +1,49 @@
+VERSION=$1
+DIR=$2
+
+REPO=https://exist-db.org/exist/apps/public-repo/public
```
This in turn should be made available as a build ARG in the Dockerfile, so we won't need to update this script when the URL changes and we have all parameters in one place.

peterstadler commented on update-xars.sh:
```diff
@@ -0,0 +1,49 @@
+
+REPO=https://exist-db.org/exist/apps/public-repo/public
+
+if [ "$DIR" != "" ]
+then
+    mkdir -p "$DIR"
+else
+    echo DIR not provided as second argument
+    exit;
+fi
```
this is not necessary if the script creates the temp folder itself (see above)

peterstadler reviewed (changes requested):
Thank you very much for the PR!
Indeed, it's a great idea to fetch the latest XARs on image build and this should be added. In addition to the minor issues I added as comments, a major improvement would be to provide the list of XARs as a build ARG in the Dockerfile and then iterate over this list in the `update-xars` script. Do you think that might be possible?

bwbohl commented:
accidentally closed my tab before submitting a longish comment: wrap up:

* make `update-xars.sh` self sufficient so it can be run standoff in order to prepare a folder that can be added to `docker runs auto deploy folder: `docker run … -v PATH:/opt/exist/autodeploy
* alternatively add a build param `$ADD_XARS_DIR` that would be added to auto deploy, too if you want to add some custom XARs.
* if the definition of which XARs to deploy should be submitted via a list of e.g. XAR-names (and then fetched by script from repo):
  * what happens if no list is provided; this might give  hint on what to get for auto deploy when running `update-xars.sh` standoff: https://github.com/eXist-db/exist/blob/5bc6a35fd6d4a91742637aefe4ad53c2324284ee/exist-installer/src/main/izpack/install.xml
  * shouldn't the exist-public-repo URL be the default? Actually on could rely on the technical expertise of the user in this context, no?

bwbohl committed:
update xars on build

bwbohl committed:
update xars on build

bwbohl committed:
Merge branch 'autoupdate-autodeploy' of https://github.com/baz-ga/existdb-docker into autoupdate-autodeploy

bwbohl commented:
@peterstadler please test the list of XARs is fixed for now; have to checkout how to submit an array as docker ARG for individualised XAR list.
But the main goal of updating the XARs on `docker build ` should be achieved.

bwbohl commented:
btw closes #2

bwbohl commented:
@peterstadler actually there's one more thing I'd like to change: make the scrip runnable in standalone by:
* allow $DIR to be set by $3 in calling `update-xars.sh`
 ```bash
if [[ -d "$3" ]]
then
    DIR="$3"
    CLEANUP=false
elif [[ "$3" != "" ]]
then
    DIR=`mkdir -p "$3"`
else
    DIR=`mktemp -d`
fi
```

and

```bash
# delete temporary dir
if [[ "$CLEANUP" = "false" ]]
then
    echo "XARs are located at: $DIR"
else
    echo "copy updated XARs to autodeploy folder"
    cp "$DIR"/*.xar ${EXIST_HOME}/autodeploy/
    rm -Rf "$DIR"
fi
```

Is that ok with you?

bwbohl committed:
enable standalone usage of update-xars.sh

bwbohl committed:
enable standalone usage of update-xars.sh

peterstadler commented:
Hi @bwbohl, just a reminder what we decided at our last conference call:

* provide the list of XARs as a build ARG in the Dockerfile and then iterate over this list in the `update-xars` script
* add more comments in the `update-xars` script for standalone mode

bwbohl committed:
Update update-xars.sh

add some documentation and user feedback

bwbohl committed:
Update update-xars.sh

add some documentation and user feedback

bwbohl committed:
Update update-xars.sh

check second argument

bwbohl committed:
Update update-xars.sh

check second argument

bwbohl committed:
Update update-xars.sh

indent

bwbohl committed:
Update update-xars.sh

indent

bwbohl committed:
Update update-xars.sh

update setting XAR repo URL

bwbohl committed:
Update update-xars.sh

update setting XAR repo URL

bwbohl committed:
Update update-xars.sh

update setting download folder

bwbohl committed:
Update update-xars.sh

update setting download folder

bwbohl committed:
Update update-xars.sh

update user feedback

bwbohl committed:
Update update-xars.sh

update user feedback

bwbohl committed:
Update update-xars.sh

fix fail on more than one argument

bwbohl committed:
Update update-xars.sh

fix fail on more than one argument

bwbohl committed:
Update update-xars.sh

fix fail on creating custom download directory

bwbohl committed:
Update update-xars.sh

fix fail on creating custom download directory

bwbohl committed:
Update update-xars.sh

add documentation comments

bwbohl committed:
Update update-xars.sh

add documentation comments

bwbohl committed:
Update update-xars.sh

* always create temporary download directory
* if $3 is set use as $PUBIR
* allow fourth argument if prune then delete XARs in $PUBDIR before copying

bwbohl committed:
Update update-xars.sh

* always create temporary download directory
* if $3 is set use as $PUBIR
* allow fourth argument if prune then delete XARs in $PUBDIR before copying

bwbohl committed:
Update update-xars.sh

* Add documentation
* Add help

bwbohl committed:
Update update-xars.sh

* Add documentation
* Add help

bwbohl committed:
Update update-xars.sh

* allow package names  to be submitted as argument
* reorder arguments

bwbohl committed:
Update update-xars.sh

* allow package names  to be submitted as argument
* reorder arguments

bwbohl committed:
Update update-xars.sh

* completely reworked parsing of arguments to parsing of options
* instead of relying on a specific order or $ARGS, e.g. `-v 5.2.0` can be used to pass the eXist-db version
* ALERT: documentation still out of sync

bwbohl committed:
Update update-xars.sh

* completely reworked parsing of arguments to parsing of options
* instead of relying on a specific order or $ARGS, e.g. `-v 5.2.0` can be used to pass the eXist-db version
* ALERT: documentation still out of sync

bwbohl committed:
Update update-xars.sh

* redefine bold and normal typeface shortcuts

bwbohl committed:
Update update-xars.sh

* redefine bold and normal typeface shortcuts

bwbohl committed:
Update update-xars.sh

* update documentation and help

bwbohl committed:
Update update-xars.sh

* update documentation and help

bwbohl committed:
update documentation in update-xars.sh

bwbohl committed:
update documentation in update-xars.sh

bwbohl committed:
Dockerfile: add ENV for XAR_REPO_URL and XAR_NAMES

bwbohl committed:
Dockerfile: add ENV for XAR_REPO_URL and XAR_NAMES

bwbohl committed:
Dockerfile: update call to update-xars.sh

bwbohl committed:
Dockerfile: update call to update-xars.sh

bwbohl commented:
~~most of it works but still I am struggling on how and when a user could pass their individual xar-list or repo url. @peterstadler could you spare a look and try?~~

done

bwbohl commented:
> most of it works but still I am struggling on how and when a user could pass their individual xar-list or repo url. @peterstadler could you spare a look and try?

To be clear my problem lies with docker, not with that shell-script ;-)

bwbohl commented:
@peterstadler some more considerations:

- [x] should the xar-update be available on build of the image?
- [x] should the xar-update be available on execution of the image, i.e. on-run?
- [ ] at the moment the default list (if not submitted) of XARs to update is hard-coded, a more flexible solution could be to:
  * read package-names from the submitted autodeploy folder
  * read version-specific package-names from the eXist-db code-repo
- [ ]  currently the script is missing the possibility to fetch from different repos

bwbohl commented:
>   * read package-names from the submitted autodeploy folder

* not possible as XAR-filenames do not necessarily match the package's short name which is used to fetch the XARs

bwbohl commented:
* should an error be reported if a submitted package nam cannot be retrieved?

bwbohl committed:
Update update-xars.sh

* for each XAR download check if file downloaded successfully, else exit

bwbohl committed:
Update update-xars.sh

* for each XAR download check if file downloaded successfully, else exit

bwbohl commented:
> * should an error be reported if a submitted package nam cannot be retrieved?

fixed in 72a310e9dc0da04bd722b84bcb445381a1771e7b

bwbohl committed:
fix XAR_REPO_URL and XAR_NAMES build args

bwbohl committed:
fix XAR_REPO_URL and XAR_NAMES build args

bwbohl commented:
> Thank you very much for the PR!
> Indeed, it's a great idea to fetch the latest XARs on image build and this should be added. In addition to the minor issues I added as comments, a major improvement would be to provide the list of XARs as a build ARG in the Dockerfile and then iterate over this list in the `update-xars` script. Do you think that might be possible?

done

bwbohl committed:
Update README.md

remove some trailing whitespaces and unnecessary line breaks

bwbohl committed:
Update README.md

remove some trailing whitespaces and unnecessary line breaks

bwbohl committed:
Update update-xars.sh

* add some documentation

bwbohl committed:
Update update-xars.sh

* add some documentation

bwbohl committed:
Update update-xars.sh

* add license

bwbohl committed:
Update update-xars.sh

* add license

bwbohl commented:
> * should the xar-update be available on execution of the image, i.e. on-run?

basically it is as you could run the `update-xars.sh` standalone and provide your local-autodeploy folder as volume option on `docker run`, e.g.

```bash
docker run --rm -it \
    -p 8080:8080 \
    --name existdb \
    -v /your/path/to/exist-data:/opt/exist/data \
    -v /your/path/to/exist-logs:/opt/exist/logs \
    -v /your/path/to/autodeploy-folder:/opt/exist/autodeploy \
    -e EXIST_ENV=development \
    -e EXIST_CONTEXT_PATH=/exist \
    stadlerpeter/existdb:latest
```

bwbohl commented:
> * [ ]   currently the script is missing the possibility to fetch from different repos

probably this could be added later

bwbohl committed:
Merge branch 'autoupdate-autodeploy' of https://github.com/baz-ga/existdb-docker into autoupdate-autodeploy

bwbohl commented:
everything's works fine so far but as the base-image has changed will have to do some update work and obviously with the new default packages for exist v5.3.0 we definitely need to get the default list from the exist repo (or somewhere else e.g. maintain a list here)

bwbohl committed:
Merge branch 'main' into autoupdate-autodeploy

bwbohl committed:
Update update-xars.sh

linting

bwbohl committed:
dynamically populate $XAR_LIST based on exist $VERSION

before this commit, there was a default list for older versions of eXist-db; now the list gets populated based on the install.xml in the exist-db repo

bwbohl commented:
@peterstadler finally I found the time to implement a dynamic population of the default XAR_LIST based on the eXist-db $VERSION and `https://raw.githubusercontent.com/eXist-db/exist/eXist-"$VERSION"/exist-installer/src/main/izpack/install.xml`

bwbohl commented:
btw: only adds functx if it is in the default deploys of the eXist-db version; so it closes #3 but functx might be missing for older db version that did not include in the default deploys

bwbohl commented:
ping

peterstadler commented:
Hi @bwbohl, sorry for the delay … 
I was just having a look and trying it out: Seems we need to install `libxml2-utils` (for `xmllint`) and some other dependencies. So, trying to keep the image size small, I thought of downloading the xars in an initial downloader step in the Dockerfile. We'd then copy over only the xars, not the tools; or do you need the script `update-xars.sh` in a running container?
If you're ok, I'd try to refactor this to make it work.

bwbohl commented:
yeah, you're probably right. The way you describe is just the way I use it e.g. in https://git.uni-paderborn.de/baz-ga/docker-website/-/blob/main/Dockerfile

Should I create a docker-xar-download (or similar) Docker image that you could just use as FROM in your Dockerfile?

bwbohl committed:
Merge branch 'peterstadler:main' into autoupdate-autodeploy
